src/Fusonic/Linq/Iterator/ChunkIterator.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of Fusonic-linq.
+ * https://github.com/fusonic/fusonic-linq
+ *
+ * (c) Fusonic GmbH
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Fusonic\Linq\Iterator;
+
+use Fusonic\Linq\Linq;
+use Iterator;
+
+/**
+ * Iterates over an iterator, returning Linq objects of the given chunk size.
+ */
+class ChunkIterator implements Iterator
+{
+    /**
+     * @var Iterator
+     */
+    private $iterator;
+
+    /**
+     * @var array
+     */
+    private $chunk;
+
+    /**
+     * @var int
+     */
+    private $i = 0;
+
+    /**
+     * @var int
+     */
+    private $chunkSize;
+
+    public function __construct(Iterator $iterator, $chunkSize)
+    {
+        $this->iterator = $iterator;
+        $this->chunkSize = $chunkSize;
+
+        $this->chunk = $this->getNextChunk();
```

Comment:
``$this->getNextChunk()`` should not be called within the constructor.
Otherwise we would have an eager evaluation of the first chunk, which is not expected.

The usual iterator calling flow looks like this:

foreach(["A","B"] as $item) { ... }

results in:

```
rewind  (Called once when starting the foreach)
next
valid  => true
current => "A"
key 

next
valid => true
current => "B"
key

next
valid => false
[TERMINATION because valid returned false]

```